### ISSUE TITLE
Automatically Generate HTML/PDF/EPUB for the Wiki Page

### ISSUE THREAD

**Issue by jonjamcam** [elieserdejesus/JamTaba#307]
This was @teamikl 's idea: Let's open the discussion!

> when document source are committed
> automatically generate HTML/PDF/EPUB then publish to web.

Here's the start of the full conversation https://github.com/elieserdejesus/JamTaba/issues/300#issuecomment-208493381

**Comment by jonjamcam** [comment]
>it's possible markdown text -> PDF, EPUB, HTML etc
if the content source was text, then it easy to make translations.

@teamikl I got a bit more into the markdown text here:

https://guides.github.com/features/mastering-markdown/

so it looks that images can be incorporated with no problems using markdown text (At first I thought wrongly it was text-only):

![image](https://cloud.githubusercontent.com/assets/15310433/14504871/f3094854-018c-11e6-8eb6-cad5ff5915d9.png)

So my question, do you think the converter is capable of handling embedded images using the code above?

**Comment by teamikl** [comment]
@jonjamcam 
ok thanks, we can start working here!

>  do you think the converter is capable of handling embedded images using the code above?

no problem, e.g. convert to HTML it will be `<img src="url" alt="Alt Text">`

> Can you elaborate on the "cost for origin creator" issue. 

I think my role on this topic is setup the workflow,
I know you don't understand where to start, yet. same for me.

I did not decide which tools to use, yet. still on research stage. I had met unexpected case, 
there was not markdown GitHub use different format for the wiki and issue comment.

the first task will be
- separate image resource and text resource (which drawing inside image)
  text resource is translatable and easy to update than image.

may be I should show the demo first and write the guide/tutorial. give me more time.

**Comment by jonjamcam** [comment]
>the first task will be

>separate image resource and text resource (which drawing inside image) text resource is translatable and easy to update than image.

@teamikl .You mean something like this:

![image](https://cloud.githubusercontent.com/assets/15310433/14519350/a04f28a0-01f4-11e6-91f2-0bd7fbc5a508.png)

**Comment by teamikl** [comment]
@jonjamcam 

Yes, and the text part move to text file.
this change make possible, translators translate text file only and share image file.

**Comment by jonjamcam** [comment]
Ok @teamikl I will work on this task in my own fork in the next days. I'm thinking of a design that is as easy to use as it is now, but with the text separated from the image.

**Comment by jonjamcam** [comment]
@teamikl I found this discussion about this topic exactly:

"Trying to convert a Github Wiki to an .epub ebook"

https://github.com/jgm/pandoc/issues/1857

**Comment by jonjamcam** [comment]
OK. so I did my first experiment with pandoc and the wiki page using the chapter:

`https://github.com/jonjamcam/JamTaba/wiki/Local-automute-entering-a-Room`

This page is also a markdown file:

`https://raw.githubusercontent.com/wiki/jonjamcam/JamTaba/Local-automute-entering-a-Room.md`

 so we use it as source for the epub. I used the command line:

`pandoc https://raw.githubusercontent.com/wiki/jonjamcam/JamTaba/Local-automute-entering-a-Room.md -o test2.epub`

Now the file `test2.epub` was created in my HDD. This was the result:

![image](https://cloud.githubusercontent.com/assets/15310433/14531615/7454d246-0234-11e6-91b9-1cdc6d5ee6c7.png)

As you can see the image was not translated correctly.

**Comment by jonjamcam** [comment]
Now I did a second experiment. Copy the content of:

`https://raw.githubusercontent.com/wiki/jonjamcam/JamTaba/Local-automute-entering-a-Room.md`

to a file in my computer called `test.md`. Now I used pandoc to convert to epub using `pandoc test.md -o test2.epub` and the resukt was:

![image](https://cloud.githubusercontent.com/assets/15310433/14533319/fd86d964-023a-11e6-8531-c56cc3828a32.png)

Now the image is converted properly, but the links are stil wrong.

**Comment by jonjamcam** [comment]
This repository claims it can "Convert GitHub wikilinks to Markdown links."

https://gist.github.com/bpj/daf711474eb53318b722

**Comment by teamikl** [comment]
@jonjamcam 
I saw now `git clone https://github.com/jonjamcam/Jamtaba.wiki`
then got markdown files.

the link format in wiki is ``[[page name]]`` that should be convert.

First i see that the filter approach, but that require install program to run the filter.
filter is small program but develop environment is huge.

I was looking for another approach,
I think documentation tool has own option to generate the navigation ToC(Table of Contents).
generate navigation automatically will reduce the cost of page movement.
I will do this task about generate page navigation.

here research going

- https://readthedocs.org/  Hosting for documentation (which use tool `Sphinx`)
- http://www.sphinx-doc.org/en/stable/  Documentation tool
  this support generate the Table of contents and navigation,
  and can generate several docs format.
  NOTE: we may not need hosting, cos HTML version is already on GitHub wiki.

The tool `Sphinx` target was another document format, I thought pandoc for convert though.
ReadTheDocs hosting had supported Sphinx with markdown format.

Now I am doing is, Setup Sphinx environment (with markdown extension) for local test

**Comment by teamikl** [comment]
@jonjamcam 

This is sample [HTML Preview](https://cloud.githubusercontent.com/assets/146129/14557470/7b05672c-0338-11e6-8943-415861c4734a.png) and [PDF](https://www.dropbox.com/s/99mz65olep6d5sb/JamtabaUserManual.pdf?dl=0) generated by Sphinx.

It was not a straight task as I expected.
I am now trying convert from github wiki's markdown files.

- pages missing title, that wiki page show the name as title.
  put file file name as title on the first line.
- navigation links were useful on wiki, other format the reader will have the navigation  
  perhaps, I can handle this by filter bottom of pages after the last horizontal bar.
- image were external links, I converted them into local links. (for embed in pdf)

It's not completed works yet at all, just report the temporary works result.

**Comment by elieserdejesus** [comment]
@jonjamcam and @teamikl , I checked the HTML preview and the PDF, the results are very interesting. All generated using the markdown as source, right?

**Comment by teamikl** [comment]
@elieserdejesus 
Yes, the source are `git clone https://github.com/jonjamcam/Jamtaba.wiki`

Currently, it requires custom filter script.
Because GitHub's wiki's markdown was not supported directly by tools.
and Markdown spec is not fully compatible to other doc format.
Markdown was more simple and easy. So, I met limitations to use Markdown as source.

e.g. missing :+1: such icons. it's easy to handle by custom script, but not yet.

**Comment by teamikl** [comment]
This is just temporary saved working result. removed the wikilinks navigation.

- [HTML Demo](https://www.dropbox.com/s/bh7dwhudnptzwr8/jamtaba-user-manual-html-demo.zip?dl=0)
- [Filter Script](https://gist.github.com/teamikl/6c008e566ba4ba5192b90f5157401b80)

**Comment by jonjamcam** [comment]
good! @teamikl 

So the html and pdf can be generated on demand automatically if a user chooses?

**Comment by teamikl** [comment]
@jonjamcam
they will be generated on commit (or push to github) timing. there are hooks
[Git Hook](https://git-scm.com/book/uz/v2/Customizing-Git-Git-Hooks) If the generator should be run local machine or [WebHook](https://developer.github.com/webhooks/) if server side.

e.g. ReadTheDocs(RTD later) hosting supports [webhook / github](http://read-the-docs.readthedocs.org/en/latest/webhooks.html)

I still have local tasks, not sure how construct the system yet.
- generate PDF require LaTeX system and fonts (it's huge libraries)
  - RTD generate PDF automatically, but reported failed cases for Japanese.
- convert markdown files by filter script (the script run local)

**Comment by jonjamcam** [comment]
@teamikl For the PDF I have my master in PDF so I can upload manually if needed.

>they will be generated on commit (or push to github) timing.

I'm not sure if we did commit for the manual. @elieserdejesus we had this discussion here https://github.com/elieserdejesus/JamTaba/issues/300#issuecomment-208341658

**Comment by jonjamcam** [comment]
@teamikl I found this resource for wiki -> PDF found by @Ezeer some time ago. Anyway I found out how to upload all the images and files to the wiki repository itself. I uploaded a copy of the master I use for the manual too.

**Comment by teamikl** [comment]
@jonjamcam 
thank u for that info. I think use *.md files as source.

wkhtmltopdf sample I tried.
It has unnecessary headers, this is single page though
side bar may be always on the right for all pages.
https://www.dropbox.com/s/7s9sth93g07f5bg/wk-sample.pdf?dl=0

I generate pdf by sphinx, ReadTheDocs also use the same tools.
Setup the pdf generation environment on local is required large tool chains.
I am thinking for now ReadTheDocs is easy solution.

1 problem for me now, is Japanese font supports.
I may need something extra settings, I found japanese pdf on ReadTheDocs.

**Comment by jonjamcam** [comment]
> wkhtmltopdf sample I tried.
It has unnecessary headers, this is single page though
side bar may be always on the right for all pages.

I see, not what we're looking for.

 I just finished correcting the wiki structure and files. We can host there too. All links are local now. No external image files and the manual I uploaded to test the viability. It's here at the bottom if you want to test:

https://github.com/jonjamcam/JamTaba/wiki/Jamtaba%27s-user-guide
